### ISSUE TITLE
Enabling/Disabling user does not work with Microsoft AD LDAP via Admin API/UI

### ISSUE THREAD

**Issue by krezovic** [keycloak/keycloak#31456]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

ldap

### Describe the bug

When using Microsoft Active Directory LDAP backend, a default LDAP mapper is pre-configured, namely "MSAD Account Controls"

![image](https://github.com/user-attachments/assets/61584c39-e3c8-473d-8004-b17b99e8144f)

Now, if the user is disabled in LDAP, enable operation simply does not work. The userAccountControl is never modified in LDAP - but UI does say that "the user has been saved"

![image](https://github.com/user-attachments/assets/0306803c-80b9-4ac2-8d64-1a02fa18d306)

Additionally, if creation of Users in LDAP is enabled (WRITABLE edit mode), then the user is by default created with userAccountControl with value "546", which, according to online decoder located at https://www.techjutsu.ca/uac-decoder decodes to

![image](https://github.com/user-attachments/assets/7de696ff-626d-40e2-b115-1d5b44a2f52e)

### Version

25.0.2

### Regression

- [ ] The issue is a regression

### Expected behavior

I am able to enable an already disabled user in LDAP and enable a newly created one - or create a user that's enabled.

### Actual behavior

I cannot enable newly created user and users disabled in LDAP

### How to Reproduce?

I am providing partial export of the LDAP component

```
{
  "realm": "primary",
  "components": {
    "org.keycloak.storage.UserStorageProvider": [
      {
        "id": "3d9e47d1-5cd5-4df4-a4ea-3cdab797fb4f",
        "name": "AD LDAP",
        "providerId": "ldap",
        "subComponents": {
          "org.keycloak.storage.ldap.mappers.LDAPStorageMapper": [
            {
              "id": "2957f424-b4a1-4ff9-a070-55b64f91e963",
              "name": "MSAD account controls",
              "providerId": "msad-user-account-control-mapper",
              "subComponents": {},
              "config": {
                "always.read.enabled.value.from.ldap": [
                  "true"
                ]
              }
            },
            {
              "id": "0e0a7c5f-de2f-4aa3-96b3-41905862d4f1",
              "name": "sAMAccountName",
              "providerId": "user-attribute-ldap-mapper",
              "subComponents": {},
              "config": {
                "ldap.attribute": [
                  "sAMAccountName"
                ],
                "attribute.force.default": [
                  "false"
                ],
                "is.mandatory.in.ldap": [
                  "true"
                ],
                "is.binary.attribute": [
                  "false"
                ],
                "read.only": [
                  "true"
                ],
                "always.read.value.from.ldap": [
                  "true"
                ],
                "user.model.attribute": [
                  "username"
                ]
              }
            },
            {
              "id": "30bac14e-cfe7-44c7-8ceb-f183be54a9ae",
              "name": "first name",
              "providerId": "user-attribute-ldap-mapper",
              "subComponents": {},
              "config": {
                "ldap.attribute": [
                  "givenName"
                ],
                "attribute.force.default": [
                  "false"
                ],
                "is.mandatory.in.ldap": [
                  "false"
                ],
                "is.binary.attribute": [
                  "false"
                ],
                "read.only": [
                  "false"
                ],
                "always.read.value.from.ldap": [
                  "true"
                ],
                "user.model.attribute": [
                  "firstName"
                ]
              }
            },
            {
              "id": "b9840a5e-5283-46b9-ad62-0c75f5fa1a21",
              "name": "accountExpires",
              "providerId": "hardcoded-ldap-attribute-mapper",
              "subComponents": {},
              "config": {
                "ldap.attribute.value": [
                  "0"
                ],
                "ldap.attribute.name": [
                  "accountExpires"
                ]
              }
            },
            {
              "id": "447be09e-1790-4120-bd06-86721374e78d",
              "name": "modify date",
              "providerId": "user-attribute-ldap-mapper",
              "subComponents": {},
              "config": {
                "ldap.attribute": [
                  "whenChanged"
                ],
                "is.mandatory.in.ldap": [
                  "false"
                ],
                "read.only": [
                  "true"
                ],
                "always.read.value.from.ldap": [
                  "true"
                ],
                "user.model.attribute": [
                  "modifyTimestamp"
                ]
              }
            },
            {
              "id": "952d61c4-46d1-48ae-bdd0-bdef55c6a0fb",
              "name": "username",
              "providerId": "user-attribute-ldap-mapper",
              "subComponents": {},
              "config": {
                "ldap.attribute": [
                  "cn"
                ],
                "attribute.force.default": [
                  "false"
                ],
                "is.mandatory.in.ldap": [
                  "true"
                ],
                "is.binary.attribute": [
                  "false"
                ],
                "always.read.value.from.ldap": [
                  "true"
                ],
                "read.only": [
                  "true"
                ],
                "user.model.attribute": [
                  "username"
                ]
              }
            },
            {
              "id": "e6fe0978-0846-4784-b975-26f2dcb88dde",
              "name": "last name",
              "providerId": "user-attribute-ldap-mapper",
              "subComponents": {},
              "config": {
                "ldap.attribute": [
                  "sn"
                ],
                "attribute.force.default": [
                  "false"
                ],
                "is.mandatory.in.ldap": [
                  "false"
                ],
                "is.binary.attribute": [
                  "false"
                ],
                "always.read.value.from.ldap": [
                  "true"
                ],
                "read.only": [
                  "false"
                ],
                "user.model.attribute": [
                  "lastName"
                ]
              }
            },
            {
              "id": "fdf58f39-125e-4675-a206-abe7d940711a",
              "name": "email",
              "providerId": "user-attribute-ldap-mapper",
              "subComponents": {},
              "config": {
                "ldap.attribute": [
                  "mail"
                ],
                "is.mandatory.in.ldap": [
                  "false"
                ],
                "attribute.force.default": [
                  "false"
                ],
                "is.binary.attribute": [
                  "false"
                ],
                "read.only": [
                  "false"
                ],
                "always.read.value.from.ldap": [
                  "false"
                ],
                "user.model.attribute": [
                  "email"
                ]
              }
            },
            {
              "id": "a1583f8e-0cfc-4202-a691-c382d15c7227",
              "name": "creation date",
              "providerId": "user-attribute-ldap-mapper",
              "subComponents": {},
              "config": {
                "ldap.attribute": [
                  "whenCreated"
                ],
                "attribute.force.default": [
                  "false"
                ],
                "is.mandatory.in.ldap": [
                  "false"
                ],
                "is.binary.attribute": [
                  "false"
                ],
                "read.only": [
                  "true"
                ],
                "always.read.value.from.ldap": [
                  "true"
                ],
                "user.model.attribute": [
                  "createTimestamp"
                ]
              }
            }
          ]
        },
        "config": {
          "serverPrincipal": [
            "HTTP/redacted@REDACTED"
          ],
          "pagination": [
            "false"
          ],
          "fullSyncPeriod": [
            "604800"
          ],
          "startTls": [
            "false"
          ],
          "connectionPooling": [
            "true"
          ],
          "usersDn": [
            "CN=Users,DC=REDACTED"
          ],
          "cachePolicy": [
            "DEFAULT"
          ],
          "useKerberosForPasswordAuthentication": [
            "false"
          ],
          "importEnabled": [
            "true"
          ],
          "enabled": [
            "true"
          ],
          "bindCredential": [
            "REDACTED"
          ],
          "bindDn": [
            "CN=keycloak,CN=Users,DC=REDACTED"
          ],
          "changedSyncPeriod": [
            "86400"
          ],
          "usernameLDAPAttribute": [
            "cn"
          ],
          "vendor": [
            "ad"
          ],
          "uuidLDAPAttribute": [
            "objectGUID"
          ],
          "allowKerberosAuthentication": [
            "true"
          ],
          "connectionUrl": [
            "ldaps://REDACTED"
          ],
          "syncRegistrations": [
            "true"
          ],
          "authType": [
            "simple"
          ],
          "krbPrincipalAttribute": [
            "userPrincipalName"
          ],
          "debug": [
            "true"
          ],
          "searchScope": [
            "2"
          ],
          "keyTab": [
            "/etc/keycloak.keytab"
          ],
          "useTruststoreSpi": [
            "always"
          ],
          "usePasswordModifyExtendedOp": [
            "false"
          ],
          "kerberosRealm": [
            "REDACTED"
          ],
          "trustEmail": [
            "true"
          ],
          "userObjectClasses": [
            "top, person, organizationalPerson, user"
          ],
          "rdnLDAPAttribute": [
            "cn"
          ],
          "readTimeout": [
            "0"
          ],
          "editMode": [
            "WRITABLE"
          ],
          "validatePasswordPolicy": [
            "true"
          ],
          "batchSizeForSync": [
            "1000"
          ]
        }
      }
    ]
}
```

For my testing, I use samba AD. If necessary, I can provide instructions on how to set up a server on a virtual machine.

**Comment by pedroigor** [comment]
@krezovic Please, share how you are using samba AD. Hopefully, it is not so complicated as the other options I have to test AD :)

**Comment by krezovic** [comment]
Hi @pedroigor,

Lets try the following:

My hardware is raspbian (Debian ARM64) bookworm. I have enabled backports repository to get latest samba 4.20. However, samba 4.17 present without backports repo is sufficient

```
$ cat /etc/apt/sources.list.d
...
deb http://deb.debian.org/debian bookworm-backports main contrib non-free non-free-firmware
...
```

Samba was installed using

```
sudo apt install -t bookworm-backports samba-ad-dc
```

(if not using backports, simply omit "-t bookworm-backports")

Rename the installed smb.conf file as AD provisioning fails

```
mv /etc/samba/smb.conf /etc/samba/smb.conf.bak
```

Provision a new AD

```
sudo samba-tool domain provision --use-rfc2307 --interactive
```

I have used following options:

```
Realm:  SAMDOM.NET
Domain [SAMDOM]: (just hit return to select default)
Server Role (dc, member, standalone) [dc]: (just hit return to select default)
DNS backend (SAMBA_INTERNAL, BIND9_FLATFILE, BIND9_DLZ, NONE) [SAMBA_INTERNAL]: (just hit return to select default)
DNS forwarder IP address (write 'none' to disable forwarding) [192.168.0.4]:  1.1.1.1 (alternatively, choose your own dns)
Administrator password: AStr0ngP@ssword
Retype password: AStr0ngP@ssword
```

(Re)start SAMBA AD

```
sudo systemctl restart samba-ad-dc
```

Note: Log file for AD is located at /var/log/samba/log.samba

Override machine's /etc/resolv.conf with samba DNS IP, but do use the real IP of the machine or Docker will not pass it to running containers. You may also use 127.0.0.1 in resolv.conf but pass dns manually via docker command line.

```
cat << EOF | sudo tee /etc/resolv.conf
nameserver 192.168.0.4
EOF
```

Once the DNS is functional, samdom.net, choosen above, will be resolvable

```
$ host samdom.net
samdom.net has address 192.168.0.4
```

Note: my machine has static IP noted above. Obviously, on your end it will resolve to yours.

Note 2: When you install docker, samba will start listening on docker bridge interface. To keep samba listening only on default ethernet/wlan, simply edit /etc/samba/smb.conf and append  the following just below the [global] section

```
interfaces = 127.0.0.1 192.168.0.4
```

Now, add keycloak bind user that has user admin privileges

```
sudo samba-tool user add keycloak AStr0ngP@ssword
sudo samba-tool user setexpiry --noexpiry keycloak
sudo samba-tool group addmembers "Account Operators" keycloak
```

Note: You can inspect LDAP object of the keycloak user

```
$ sudo samba-tool user show keycloak
dn: CN=keycloak,CN=Users,DC=samdom,DC=net
...
```

(we only care for DN for configuration of KC)

Lets setup Keycloak. The following section uses docker to boot keycloak on the same machine. Docker is installed as shown in official docs for Debian https://docs.docker.com/engine/install/debian/#install-using-the-repository (works on ARM64).

```
$ sudo adduser $(whoami) docker
(relogin)
$ docker version
Client: Docker Engine - Community
 Version:           27.1.0
 API version:       1.46
 Go version:        go1.21.12
 Git commit:        6312585
 Built:             Fri Jul 19 17:43:14 2024
 OS/Arch:           linux/arm64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          27.1.0
  API version:      1.46 (minimum version 1.24)
  Go version:       go1.21.12
  Git commit:       a21b1a2
  Built:            Fri Jul 19 17:43:14 2024
  OS/Arch:          linux/arm64
  Experimental:     false
 containerd:
  Version:          1.7.19
  GitCommit:        2bf793ef6dc9a18e00cb12efb64355c2c9d5eb41
 runc:
  Version:          1.7.19
  GitCommit:        v1.1.13-0-g58aa920
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

Run keycloak in dev mode, exposed via http port 8080

```
docker run -d --name keycloak-test -e KEYCLOAK_ADMIN=admin -e KEYCLOAK_ADMIN_PASSWORD=admin -p 8080:8080 quay.io/keycloak/keycloak:25.0.1 start-dev
```

I can access it via http://192.168.0.4:8080 on my end. Your end may vary.

Now, to spare myself some trouble, I will configure a new realm "ldap" which you can import on your end, provided you use same credentials as I used above. This assumes that docker and samba run on the same machine and that keycloak can resolve samdom.net domain hosted by samba dns

**Comment by krezovic** [comment]
Hi @pedroigor,

Lets try the following:

My hardware is raspbian (Debian ARM64) bookworm. I have enabled backports repository to get latest samba 4.20. However, samba 4.17 present without backports repo is sufficient

```
$ cat /etc/apt/sources.list.d
...
deb http://deb.debian.org/debian bookworm-backports main contrib non-free non-free-firmware
...
```

Samba was installed using

```
sudo apt install -t bookworm-backports samba-ad-dc
```

(if not using backports, simply omit "-t bookworm-backports")

Rename the installed smb.conf file as AD provisioning fails

```
mv /etc/samba/smb.conf /etc/samba/smb.conf.bak
```

Provision a new AD

```
sudo samba-tool domain provision --use-rfc2307 --interactive
```

I have used following options:

```
Realm:  SAMDOM.NET
Domain [SAMDOM]: (just hit return to select default)
Server Role (dc, member, standalone) [dc]: (just hit return to select default)
DNS backend (SAMBA_INTERNAL, BIND9_FLATFILE, BIND9_DLZ, NONE) [SAMBA_INTERNAL]: (just hit return to select default)
DNS forwarder IP address (write 'none' to disable forwarding) [192.168.0.4]:  1.1.1.1 (alternatively, choose your own dns)
Administrator password: AStr0ngP@ssword
Retype password: AStr0ngP@ssword
```

Edit /etc/samba/smb.conf, and add two lines just below [global] section. This makes simple bind a bit easier by not requiring us to set up TLS certs.

```
        ldap server require strong auth = no
        interfaces = 127.0.0.1 192.168.0.4
```

The final /etc/samba/smb.conf looks like

```
$ cat /etc/samba/smb.conf
# Global parameters
[global]
        ldap server require strong auth = no
        interfaces = 127.0.0.1 192.168.0.4
        dns forwarder = 1.1.1.1
        netbios name = SERVER3
        realm = SAMDOM.NET
        server role = active directory domain controller
        workgroup = SAMDOM
        idmap_ldb:use rfc2307 = yes

[sysvol]
        path = /var/lib/samba/sysvol
        read only = No

[netlogon]
        path = /var/lib/samba/sysvol/samdom.net/scripts
        read only = No
```

(Re)start SAMBA AD

```
sudo systemctl restart samba-ad-dc
```

Note: Log file for AD is located at /var/log/samba/log.samba

Override machine's /etc/resolv.conf with samba DNS IP, but do use the real IP of the machine or Docker will not pass it to running containers. You may also use 127.0.0.1 in resolv.conf but pass dns manually via docker command line.

```
cat << EOF | sudo tee /etc/resolv.conf
nameserver 192.168.0.4
EOF
```

Once the DNS is functional, samdom.net, choosen above, will be resolvable

```
$ host samdom.net
samdom.net has address 192.168.0.4
```

Note: my machine has static IP noted above. Obviously, on your end it will resolve to yours.

Now, add keycloak bind user that has user admin privileges

```
sudo samba-tool user add keycloak AStr0ngP@ssword
sudo samba-tool user setexpiry --noexpiry keycloak
sudo samba-tool group addmembers "Account Operators" keycloak
```

Note: You can inspect LDAP object of the keycloak user

```
$ sudo samba-tool user show keycloak
dn: CN=keycloak,CN=Users,DC=samdom,DC=net
...
```

(we only care for DN for configuration of KC)

Lets setup Keycloak. The following section uses docker to boot keycloak on the same machine. Docker is installed as shown in official docs for Debian https://docs.docker.com/engine/install/debian/#install-using-the-repository (works on ARM64).

```
$ sudo adduser $(whoami) docker
(relogin)
$ docker version
Client: Docker Engine - Community
 Version:           27.1.0
 API version:       1.46
 Go version:        go1.21.12
 Git commit:        6312585
 Built:             Fri Jul 19 17:43:14 2024
 OS/Arch:           linux/arm64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          27.1.0
  API version:      1.46 (minimum version 1.24)
  Go version:       go1.21.12
  Git commit:       a21b1a2
  Built:            Fri Jul 19 17:43:14 2024
  OS/Arch:          linux/arm64
  Experimental:     false
 containerd:
  Version:          1.7.19
  GitCommit:        2bf793ef6dc9a18e00cb12efb64355c2c9d5eb41
 runc:
  Version:          1.7.19
  GitCommit:        v1.1.13-0-g58aa920
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

Run keycloak in dev mode, exposed via http port 8080

```
$ docker run -d --add-host samdom.net:192.168.0.4 --dns 1.1.1.1 --name keycloak-test -e KEYCLOAK_ADMIN=admin -e KEYCLOAK_ADMIN_PASSWORD=admin -p 8080:8080 quay.io/keycloak/keycloak:25.0.2 start-dev
```

I can access it via http://192.168.0.4:8080 on my end. Your end may vary.

Now, to spare myself some trouble, I will configure a new realm "ldap" which you can import on your end, provided you use same credentials as I used above. This assumes that docker and samba run on the same machine and that keycloak can resolve samdom.net domain hosted by samba dns. I have opted for using --add-host as shown above since I seem to be having troubles with IPv6. A simple test should work

```
$ docker exec -it keycloak-test bash -c "getent hosts samdom.net"
192.168.0.4     samdom.net
```

The configuration is quite straightforward and uses defaults except one additional mapper which maps keycloak username to Samba AD "sAMAccountName" so user creation works.

Once realm import is done, you may navigate to ldap realm, Users and search for "keycloak" - the user the bind is done with. It should return it

You may create any user. It will be reflected in samba ad and may be inspected with "sudo samba-tool user show <username>". For testing, I have created user "test", simply by only typing one required attribute into Keycloak "Add User" form - the username. It is shown in samba as follows

```
$ sudo samba-tool user show test
dn: CN=test,CN=Users,DC=samdom,DC=net
objectClass: top
objectClass: person
objectClass: organizationalPerson
objectClass: user
cn: test
sn:
givenName:
instanceType: 4
whenCreated: 20240723192523.0Z
whenChanged: 20240723192523.0Z
uSNCreated: 4293
uSNChanged: 4293
name: test
objectGUID: 30a5b2a9-2968-41c1-8d3d-9333b32bf881
userAccountControl: 546
badPwdCount: 0
codePage: 0
countryCode: 0
badPasswordTime: 0
lastLogoff: 0
lastLogon: 0
pwdLastSet: 0
primaryGroupID: 513
objectSid: S-1-5-21-306620582-3174703427-1138850931-1104
accountExpires: 9223372036854775807
logonCount: 0
sAMAccountName: test
sAMAccountType: 805306368
objectCategory: CN=Person,CN=Schema,CN=Configuration,DC=samdom,DC=net
distinguishedName: CN=test,CN=Users,DC=samdom,DC=net
```

Obviously, it is shown as "Disabled" in Keycloak. You may enable it via samba-tool

```
$ sudo samba-tool user enable test
```

Refresh of Keycloak user UI will reflect the attribute change.

Now, the realm JSON is below:

```
{
  "realm": "ldap",
  "components": {
    "org.keycloak.storage.UserStorageProvider": [
      {
        "name": "ldap",
        "providerId": "ldap",
        "subComponents": {
          "org.keycloak.storage.ldap.mappers.LDAPStorageMapper": [
            {
              "name": "modify date",
              "providerId": "user-attribute-ldap-mapper",
              "subComponents": {},
              "config": {
                "ldap.attribute": [
                  "whenChanged"
                ],
                "is.mandatory.in.ldap": [
                  "false"
                ],
                "read.only": [
                  "true"
                ],
                "always.read.value.from.ldap": [
                  "true"
                ],
                "user.model.attribute": [
                  "modifyTimestamp"
                ]
              }
            },
            {
              "name": "last name",
              "providerId": "user-attribute-ldap-mapper",
              "subComponents": {},
              "config": {
                "ldap.attribute": [
                  "sn"
                ],
                "is.mandatory.in.ldap": [
                  "true"
                ],
                "read.only": [
                  "false"
                ],
                "always.read.value.from.ldap": [
                  "true"
                ],
                "user.model.attribute": [
                  "lastName"
                ]
              }
            },
            {
              "name": "MSAD account controls",
              "providerId": "msad-user-account-control-mapper",
              "subComponents": {},
              "config": {
                "always.read.enabled.value.from.ldap": [
                  "true"
                ]
              }
            },
            {
              "name": "username",
              "providerId": "user-attribute-ldap-mapper",
              "subComponents": {},
              "config": {
                "ldap.attribute": [
                  "cn"
                ],
                "is.mandatory.in.ldap": [
                  "true"
                ],
                "read.only": [
                  "false"
                ],
                "always.read.value.from.ldap": [
                  "false"
                ],
                "user.model.attribute": [
                  "username"
                ]
              }
            },
            {
              "name": "email",
              "providerId": "user-attribute-ldap-mapper",
              "subComponents": {},
              "config": {
                "ldap.attribute": [
                  "mail"
                ],
                "is.mandatory.in.ldap": [
                  "false"
                ],
                "read.only": [
                  "false"
                ],
                "always.read.value.from.ldap": [
                  "false"
                ],
                "user.model.attribute": [
                  "email"
                ]
              }
            },
            {
              "name": "creation date",
              "providerId": "user-attribute-ldap-mapper",
              "subComponents": {},
              "config": {
                "ldap.attribute": [
                  "whenCreated"
                ],
                "is.mandatory.in.ldap": [
                  "false"
                ],
                "always.read.value.from.ldap": [
                  "true"
                ],
                "read.only": [
                  "true"
                ],
                "user.model.attribute": [
                  "createTimestamp"
                ]
              }
            },
            {
              "name": "sAMAccountName",
              "providerId": "user-attribute-ldap-mapper",
              "subComponents": {},
              "config": {
                "ldap.attribute": [
                  "sAMAccountName"
                ],
                "is.mandatory.in.ldap": [
                  "true"
                ],
                "attribute.force.default": [
                  "true"
                ],
                "is.binary.attribute": [
                  "false"
                ],
                "always.read.value.from.ldap": [
                  "true"
                ],
                "read.only": [
                  "true"
                ],
                "user.model.attribute": [
                  "username"
                ]
              }
            },
            {
              "name": "first name",
              "providerId": "user-attribute-ldap-mapper",
              "subComponents": {},
              "config": {
                "ldap.attribute": [
                  "givenName"
                ],
                "is.mandatory.in.ldap": [
                  "true"
                ],
                "always.read.value.from.ldap": [
                  "true"
                ],
                "read.only": [
                  "false"
                ],
                "user.model.attribute": [
                  "firstName"
                ]
              }
            }
          ]
        },
        "config": {
          "fullSyncPeriod": [
            "-1"
          ],
          "pagination": [
            "false"
          ],
          "startTls": [
            "false"
          ],
          "connectionPooling": [
            "false"
          ],
          "usersDn": [
            "CN=Users,DC=samdom,DC=net"
          ],
          "cachePolicy": [
            "DEFAULT"
          ],
          "useKerberosForPasswordAuthentication": [
            "false"
          ],
          "importEnabled": [
            "true"
          ],
          "enabled": [
            "true"
          ],
          "changedSyncPeriod": [
            "-1"
          ],
          "bindDn": [
            "CN=keycloak,CN=Users,DC=samdom,DC=net"
          ],
          "usernameLDAPAttribute": [
            "cn"
          ],
          "bindCredential": [
            "AStr0ngP@ssword"
          ],
          "vendor": [
            "ad"
          ],
          "uuidLDAPAttribute": [
            "objectGUID"
          ],
          "connectionUrl": [
            "ldap://samdom.net"
          ],
          "allowKerberosAuthentication": [
            "false"
          ],
          "syncRegistrations": [
            "true"
          ],
          "authType": [
            "simple"
          ],
          "krbPrincipalAttribute": [
            "userPrincipalName"
          ],
          "searchScope": [
            "1"
          ],
          "useTruststoreSpi": [
            "always"
          ],
          "usePasswordModifyExtendedOp": [
            "false"
          ],
          "trustEmail": [
            "false"
          ],
          "userObjectClasses": [
            "person, organizationalPerson, user"
          ],
          "rdnLDAPAttribute": [
            "cn"
          ],
          "editMode": [
            "WRITABLE"
          ],
          "validatePasswordPolicy": [
            "false"
          ]
        }
      }
    ]
  }
 }
```

**Comment by pedroigor** [comment]
@krezovic Thanks a lot. I think I see what is happening.

The `ACCOUNTDISABLED` control is only set if the `pwdLastSet` (date/time password was set/updated) is set. When creating a user, the user is disabled because of that. No matter what you do, you won't be able to enable the user while no password is set for the user.

In regards to setting the `ACCOUNTDISABLED` in AD, it will work after syncing users. But there is an issue when propagating `ENABLED` to LDAP because the control is not sent to LDAP. I'm sending a PR with a fix.

**Comment by pedroigor** [comment]
@krezovic Would you mind checking https://github.com/keycloak/keycloak/pull/31554?

**Comment by krezovic** [comment]
I hope I have downloaded the correct tarball (the dist one from here https://github.com/keycloak/keycloak/actions/runs/10067141152?pr=31554), but the issue still persists. Enable button does not work

![image](https://github.com/user-attachments/assets/26107726-af3b-4d4e-a9eb-e42f1638ed87)

I did notice that when setting password using "Set Password" tab sets the userAccountControl to 514 (clears the PASSWD_NOTREQD flag). However, pwdLastSet is not updated at all and if I try to log in via kerberos, I get the password change prompt:

![image](https://github.com/user-attachments/assets/6871edd4-f969-4cc4-b557-f47493534861)

**Comment by krezovic** [comment]
I am not an expert, but this line looks wrong

https://github.com/keycloak/keycloak/blob/b9dae2a2ff3a6f0a27586bb5574745680e360bb2/federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/msad/MSADUserAccountControlStorageMapper.java#L102

The user is already disabled due to flag being set and the code is supposed to enable it - should it not be inverted?

Also, pwdLastSet is not propagated to ldap most likely due to -1 being invalid value - prompting in password reset when trying to log in via kerberos

https://github.com/keycloak/keycloak/blob/b9dae2a2ff3a6f0a27586bb5574745680e360bb2/federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/msad/MSADUserAccountControlStorageMapper.java#L96

**Comment by krezovic** [comment]
Please disregard my comment. I was setting password with "Temporary" flag set. Once I deselected "Temporary" flag, the pwdLastSet and userAccountControl can get updated. It is a bit of a chicken and egg problem, set password to non-temporary to enable user so they can log in, then you have to add manual action to reset password. Not really friendly when you want to support both AD and Keycloak.

**Comment by pedroigor** [comment]
@krezovic The problem with removing that line is that we cannot assume that a password update should enable the user.

The problem you described is what is actually happening and I agree the UX is not the best. Basically, setting a temporary password will set the `pwdLastSet` attribute to `0` and prevent the administrator from enabling/disabling users until they don't set a non-temporary password.

I'm not sure if we can change the behavior now and remove the `pwdLastSet` check [1] when enabling/disabling LDAP users. It will change the behavior and for that we need more discussions to understand the motivation about only updating the `ACCOUNTDISABLED` user control if the `pwdLastSet` is set.

For this reason, I'm moving this issue to our backlog as we should be looking at other UX issues related to LDAP.

[1] https://github.com/keycloak/keycloak/blob/main/federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/msad/MSADUserAccountControlStorageMapper.java#L251

**Comment by pedroigor** [comment]
There are different behaviors when using AD where we can improve:

#### Enabled/Disabled User when creating via Admin API

When the user is created via Admin API (e.g.: admin console), the account will created in both the local database and AD. However, in LDAP the user is not yet set with a credential and, as such, the account is marked as disabled by default.

This particular behavior is what is described here so that trying to enable the user via Admin API will return a successful operation but behind the scenes, it did nothing because the user is still disabled in LDAP and in the local database.

For this specific issue, we should return a proper error message saying that the account cannot be enabled (or updated) until a non-temporary password is set to the account and propagated to the LDAP in order to make it possible to enable the account in both LDAP and Keycloak sides.

We should probably update the documentation to make this behavior more clear.

#### Error when enabling the user after setting a password for the first time

When the user is created via Admin API (e.g.: admin console), the account will created in both the local database and AD. However, in LDAP the user is not yet set with a credential and, as such, the account is marked as disabled by default.

The admin might want to set a password to the user (in LDAP) so that he can enable the user and ask the user to change the password through the update password required action.

If you just created the user and you try to set the first password for the user, you won't be able to enable the user once the password is set. That is because the UI is not fetching the up-to-date state of the user after successfuly setting the password so trying to enable or just update the user will cause a validation error because of the value of the `modifiedTimestamp`  attribute that is different than what is on the server.

For this specific issue, we probably want to re-render the UI so that the user representation is up-to-date and you can perform operations like update or enable the user.

#### Required action `Update password` set to newly created users

When the user is created via Admin API (e.g.: admin console), the account will created in both the local database and AD. However, in LDAP the user is not yet set with a credential and, as such, the account is marked as disabled by default.

In this case, the account, even though disabled, is created and automatically set with a `Update Password` required action. Adding this required action does not make sense if the user is disabled because he won't be able to log in until a password is set by the administrator (and updated in LDAP) and the user is enabled.

For this specific issue, we probably want to not add the `Update password` required action but just let administrators to add the action by themselves.

This is also an issue when the LDAP provider is not importing users because you cannot update or perform any action on a user because required actions can not be set to federated/external users.

**Comment by sguilhen** [comment]
> For this specific issue, we should return a proper error message saying that the account cannot be enabled (or updated) until a non-temporary password is set to the account and propagated to the LDAP in order to make it possible to enable the account in both LDAP and Keycloak sides.

Agreed, if we can somehow detect this in code it would be nice to have a better message displayed.

> For this specific issue, we probably want to re-render the UI so that the user representation is up-to-date and you can perform operations like update or enable the user.

Yeah, users should not get an error in this situation

> For this specific issue, we probably want to not add the `Update password` required action but just let administrators to add the action by themselves.

You've raised a good point for required actions in general.. if user is disabled, it doesn't make sense to add a required action because the user can't possibly login in the first place, let alone follow a required action afterward.
